src/redux/slices/user.js

```diff
@@ -213,27 +216,107 @@ export const transferNFT = createAsyncThunk(
   }
 );
 
+export const getContacts = createAsyncThunk(
+  'keyring/getContacts',
+  async (walletNumber = 0, { getState, dispatch }) => {
+    try {
+      dispatch(setContactsLoading(true));
+      const state = getState();
+      const res = await state.keyring.instance?.getContacts(walletNumber);
+      return res?.map(formatContact);
+    } catch (e) {
+      dispatch(setContactsLoading(false));
+      console.log('Error getting contacts:', e);
+    }
+  }
+);
+
+export const addContact = createAsyncThunk(
+  'keyring/addContact',
+  async ({ contact, walletNumber = 0, onFinish }, { getState, dispatch }) => {
+    try {
+      dispatch(setContactsLoading(true));
+      const state = getState();
+      const res = await state.keyring.instance?.addContact(
+        formatContactForController(contact),
+        walletNumber
+      );
+      if (res) {
+        dispatch(getContacts())
+          .unwrap()
+          .then(() => onFinish?.());
+      }
+    } catch (e) {
+      // TODO: We should handle this error
+      dispatch(setContactsLoading(false));
+      console.log('Error adding contacts:', e);
+    }
+  }
+);
+
+export const removeContact = createAsyncThunk(
+  'keyring/removeContact',
+  async ({ contactName, walletNumber = 0 }, { getState, dispatch }) => {
+    try {
+      dispatch(setContactsLoading(true));
+      const state = getState();
+      const res = await state.keyring.instance?.deleteContact(
+        contactName,
+        walletNumber
+      );
+      if (res) {
+        dispatch(getContacts());
+      }
+    } catch (e) {
+      // TODO: We should handle this error
+      dispatch(setContactsLoading(false));
+      console.log('Error removing contact:', e);
+    }
+  }
+);
+
+export const editContact = createAsyncThunk(
+  'keyring/editContact',
+  async ({ contact, newContact, walletNumber = 0 }, { getState, dispatch }) => {
+    try {
+      dispatch(setContactsLoading(true));
+      const state = getState();
+      const removeContactRes = await state.keyring.instance?.deleteContact(
+        contact.name,
+        walletNumber
+      );
+      const addContactRes = await state.keyring.instance?.addContact(
+        formatContactForController(newContact),
+        walletNumber
+      );
+      if (removeContactRes && addContactRes) {
+        dispatch(getContacts());
```

Comment:
same here! Also, what happens if the contact is deleted and the addContact fails? thats something we should try to handle

Reply:
same answer. Yes i agree that's the only weird case, it shouldnt happen that thats the border case that we have, the thing is that dab doesnt have a edit method to handle it